docs/_docs/05_rfc/variable_output_modification.md

```diff
@@ -0,0 +1,152 @@
+  [consul-security-group-rules
+  module](https://registry.terraform.io/modules/hashicorp/consul/aws/latest/submodules/consul-security-group-rules)
+
+At its core, Terragrunt is designed to add the necessary ingredients to convert **Shared Service Modules** into **Root
+Modules** so that they can be deployed with `terraform apply`. Features like `generate` and `remote_state` support this
+transition by injecting the necessary blocks that support directly invoking `terraform`.
+
+Note that the distinction between **Shared Service Modules** and **Shared Modules** is subtle, and oftentimes is not a
+clear cut technical difference. However, there are technical limitations in the current Terragrunt implementation that
+prevents deploying certain shared modules:
+
+- Every complex input must have a `type` associated with it. Otherwise, Terraform will interpret the input that
+  Terragrunt passes through as `string`. This includes `list` and `map`.
+- Derived sensitive outputs must be marked as `sensitive`. Refer to the [terraform tutorial on sensitive
+  variables](https://learn.hashicorp.com/tutorials/terraform/sensitive-variables#reference-sensitive-variables) for more
+  information on this requirement.
+
+Terraform only enforces these restrictions on the **Root Module**. This means that there are some shared modules on the
+registry that Terragrunt can not deploy. Note that the lack of these parameters may not by itself indicate that they are
+**Shared Modules**. That is, these modules may be **Shared Service Modules** by design, but because they are
+only designed for use with Terraform, they may not set the `type` or `sensitive` flags on the `variable` and `output`
+blocks, preventing usage as a transformed root module unless those inputs and outputs are configured.
+
+
+## Proposed solution
+
+To handle this, this RFC proposes a new block: `transform`. Here is an example:
+
+Consider a module that has the following:
+
+```hcl
+variable "my_password" {
+  sensitive = true
+}
+
+variable "my_list" {}
+
+# NOTE: this must be marked as sensitive since it is derived from a sensitive variable
+output "my_password_hashed" {
+  value = base64sha256(var.my_password)
+}
+
+output "length_my_list" {
+  value = length(var.my_list)
+}
+```
+
+Using Terragrunt with this module will run into the following issue:
+
+- Because the output `my_password_hashed` is not marked as sensitive, terraform will error out.
+- `my_list` is missing the type definition, so the input from `terragrunt` will be interpretted as a string. This means
+  that the output `length_my_list` will be the string length, and not the list length.
+
+We will need to transform these variables and outputs. We will introduce the `transform` block to handle this. The
+following `terragrunt.hcl` configuration indicates the necessary transformations to `variable` and `output` to support
+deployment:
+
+```hcl
+transform {
```

Comment:
Would this be an attribute of the `terraform` block, since it would be specifically related to the `source` argument?

Reply:
Theoretically, this could also rewrite resources? 🤯

Reply:
I kind of like it being a separate block, given that the `terraform` block is more about how terragrunt calls terraform, and transforming the code is a different operation. In that regard, it is more similar to `generate` than `terraform`.

Reply:
Hmm, and that makes me wonder what kind of terragrunt config/feature might it be to support multiple terraform blocks...